pkg/cmd/server/middleware.go

```diff
@@ -0,0 +1,207 @@
+package server
+
+import (
+	"fmt"
+
+	"github.com/authzed/spicedb/pkg/util"
+
+	"google.golang.org/grpc"
+)
+
+// MiddlewareChain describes an ordered sequence of middlewares that can be modified
+// with one or more MiddlewareModification. This struct is used to facilitate the
+// creation and modification of gRPC middleware chains
+type MiddlewareChain struct {
+	chain []ReferenceableMiddleware
+}
+
+// NewMiddlewareChain creates a new middleware chain given a zero or more named middlewares.
+// An error will be returned in case validation of the NamedMiddlewares fail.
+func NewMiddlewareChain(mw ...ReferenceableMiddleware) (MiddlewareChain, error) {
+	if err := validate(mw); err != nil {
+		return MiddlewareChain{}, err
+	}
+	return MiddlewareChain{chain: mw}, nil
+}
+
+// MiddlewareModification describes an operation to modify a MiddlewareChain
+type MiddlewareModification struct {
+	// Dependency is used to define with respect to which middleware an operation is performed
+	Dependency string
+	// Operation describes the type of operation to be performed
+	Operation MiddlewareOperation
+	// Middlewares the named middlewares that will be part of this modification
+	Middlewares []ReferenceableMiddleware
+}
+
+func (mm MiddlewareModification) validate() error {
+	if mm.Operation != OperationReplaceAllUnsafe && mm.Dependency == "" {
+		return fmt.Errorf("cannot perform middleware modification without a dependency: %v", mm)
+	}
+	if err := validate(mm.Middlewares); err != nil {
+		return err
+	}
+	return nil
+}
+
+func validate(mws []ReferenceableMiddleware) error {
+	names := util.NewSet[string]()
+	for _, mw := range mws {
+		if mw.Name == "" {
+			return fmt.Errorf("unnamed middleware found: %v", mw)
+		}
+		if !names.Add(mw.Name) {
+			return fmt.Errorf("found middleware with duplicate names in middleware modification: %s", mw.Name)
+		}
+	}
+	return nil
+}
+
+// ReferenceableMiddleware represents a middleware in a MiddlewareChain. Middlewares can
+// be referenceable in MiddlewareModification, for example "append after middleware abc".
+// Middlewares are referenced by name.
+// A named middleware can be also be unreferenceable, in which case can be left unnamed.
```

Comment:
in which case the middleware can be left...

Reply:
this is actually stale. I changed the semantics here a bit: ReferenceableMiddleware can now be marked as `internal`. This only means you cannot replace that middleware, but can reference it for append or prepend. The goal is to mark middleware that shouldn't be removed at all.